engine/src/main/java/org/pentaho/di/trans/steps/userdefinedjavaclass/UserDefinedJavaClassMeta.java

```diff
@@ -139,9 +136,37 @@ public UserDefinedJavaClassMeta() {
       "org.pentaho.di.trans.steps.userdefinedjavaclass.*", "org.pentaho.di.trans.step.*",
       "org.pentaho.di.core.row.*", "org.pentaho.di.core.*", "org.pentaho.di.core.exception.*" } );
 
-    cbe.cook( new Scanner( null, sr ) );
+    Scanner scanner = new Scanner( null, sr );
 
-    return cbe.getClazz();
+    //run cook in separate thread for avoid crash the application in case of an unchecked compilation error
+    final boolean[] finishedCorrect = { false };
+    final Exception[] checkedException = new Exception[ 1 ];
+
+    Runnable runnable = () -> {
+      try {
+        cbe.cook( scanner );
+        finishedCorrect[ 0 ] = true;
+      } catch ( Exception e ) {
+        checkedException[ 0 ] = e;
+      }
+    };
+
+    Thread thread = new Thread( runnable );
+    thread.start();
+
+    while ( true ) {
```

Comment:
If we really must compile in a separate thread, please void this synchronization code. You should use Executors/Future (or wait/notify along with a monitor object if you really feel like writing synchronization code)